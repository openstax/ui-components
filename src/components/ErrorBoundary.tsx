import * as Sentry from '@sentry/react';
import React from 'react';
import { Error as ErrorComponent, ErrorPropTypes } from './Error';
import type { ErrorBoundaryProps } from '@sentry/react/types/errorboundary';
import { ErrorContext } from '../contexts';
import { SentryError } from '../types';

const Error = ({ children, ...props }: React.PropsWithChildren<ErrorPropTypes>) =>
  <ErrorComponent data-testid='error-fallback' {...props}>{children}</ErrorComponent>;

const defaultErrorFallbacks = {
  'generic': <Error data-testid='error-fallback' />,
  'SessionExpiredError': <Error heading='Your session has expired'>
    Please refresh your browser and try again. If this doesn't solve the problem, visit our <a href="https://openstax.secure.force.com/help" target="_blank">Support Center</a>.
  </Error>,
  'UnauthorizedError': <Error heading="Uh-oh, it seems you can't access this page.">
    You may not have the required permissions or may have been logged out. Try refreshing the page or restarting your browser.
    If the issue persists, visit our <a href="https://openstax.secure.force.com/help" target="_blank">Support Center</a>.
  </Error>
};

export const getTypeFromError = (error: Error | PromiseRejectionEvent['reason']) => {
  if (!error) { return undefined };
  const { TYPE, name } = error.constructor;
  return TYPE && typeof TYPE === 'string' ? TYPE : name;
};

export const ErrorBoundary = ({
  children,
  renderFallback,
  fallback = defaultErrorFallbacks['generic'],
  catchUnhandledRejections = true,
  windowImpl = window,
  sentryDsn,
  sentryInit,
  ...props
}: ErrorBoundaryProps & {
  renderFallback?: boolean;
  catchUnhandledRejections?: boolean;
  windowImpl?: Window | Pick<Window, 'addEventListener' | 'removeEventListener'>;
  sentryDsn?: string;
  sentryInit?: Sentry.BrowserOptions;
  errorFallbacks?: { [_: string]: JSX.Element }
}) => {
  const [error, setError] = React.useState<SentryError & { type?: string } | null>(null);
  const errorFallbacks = { ...defaultErrorFallbacks, ...props.errorFallbacks };
  const typedFallback = error?.type ? errorFallbacks[error.type] : undefined;

  // Optionally re-render with the children so they can display inline errors with <ErrorMessage />
  const renderElement = error && renderFallback ? (typedFallback || fallback) : <>{children}</>;

  React.useEffect(() => {
    if (!sentryDsn && !sentryInit) {
      return;
    }

    Sentry.init(sentryInit || {
      dsn: sentryDsn,
      environment: window.location.hostname,
      integrations: [
        new Sentry.BrowserTracing(),
      ],
    });
  }, []);

  React.useEffect(() => {
    if (!catchUnhandledRejections) {
      return;
    }

    const handleRejection = (e: PromiseRejectionEvent) => {
      setError({
        error: {
          name: e.type,
          message: e.reason?.toString(),
        },
        type: getTypeFromError(e.reason),
      });
    };

    windowImpl.addEventListener('unhandledrejection', handleRejection);
    return () => windowImpl.removeEventListener('unhandledrejection', handleRejection);
  }, []);
  // There are two references to the render element here because the Sentry fallback (and onError)
  // are not used for unhandledrejection events. To support those events, we add a listener and
  // reuse the same error state and rendering logic.
  return <ErrorContext.Provider value={error}>
    <Sentry.ErrorBoundary
      fallback={renderElement}
      onError={(error, componentStack, eventId) => {
        setError({
          error,
          // If the error is a custom error from ts-utils, use the custom type instead of 'Error'
          type: getTypeFromError(error),
          componentStack,
          eventId
        });
      }}
      {...props}
      onReset={() => setError(null)}
    >
      {renderElement}
    </Sentry.ErrorBoundary>
  </ErrorContext.Provider>;
}
