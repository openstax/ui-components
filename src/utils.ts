export const assertNotNull = <T>(obj: T | null, message: string) => {
  if (obj !== null) { return obj; }
  throw new Error(message);
};

export const focusElement = (element?: Element | null) => {
  if (element instanceof HTMLElement) { element.focus(); }
};

export const getTypeFromError = (error: Error | PromiseRejectionEvent['reason']) => {
  if (!error) { return undefined; }
  const { TYPE, name } = error.constructor;
  return TYPE && typeof TYPE === 'string' ? TYPE : name;
};
