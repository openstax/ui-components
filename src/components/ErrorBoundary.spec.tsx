import renderer, { ReactTestRenderer, act } from 'react-test-renderer';
import { ErrorBoundary } from './ErrorBoundary';
import sentryTestkit from 'sentry-testkit';
import * as Sentry from '@sentry/react';
import { findByTestId } from '../test/utils';
import { SessionExpiredError } from '@openstax/ts-utils/errors';

const { testkit, sentryTransport } = sentryTestkit();

const ErrorComponent = () => { throw new Error('Test Error') };

describe('ErrorBoundary', () => {
  beforeAll(() => {
    Sentry.init({
      dsn: 'https://examplePublicKey@o0.ingest.sentry.io/0',
      transport: sentryTransport,
    });

    jest.spyOn(Sentry, 'lastEventId').mockReturnValue('someuuid');
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('renders children', () => {
    const tree = renderer.create(
      <ErrorBoundary>
        Test
      </ErrorBoundary>
    ).toJSON();
    expect(tree).toMatchInlineSnapshot(`"Test"`);
  });

  it('captures errors', () => {
    const spy = jest.spyOn(console, 'error')
    spy.mockImplementation(() => undefined);

    const render = renderer.create(
      <ErrorBoundary renderFallback>
        <ErrorComponent />
      </ErrorBoundary>
    );

    expect(findByTestId(render.root, 'error-fallback')).toBeTruthy();
    expect(testkit.reports()).toHaveLength(1);

    spy.mockRestore();
  });

  it('resets error', () => {
    const spy = jest.spyOn(console, 'error')
    spy.mockImplementation(() => undefined);

    let render: ReactTestRenderer;
    expect(() => {
      render = renderer.create(
        <ErrorBoundary
          renderFallback
          fallback={({ resetError }: {
            resetError: () => void
          }) => { resetError(); return <></> }}
        >
          <ErrorComponent />
        </ErrorBoundary>
      );
    }).toThrow();

    expect(() => findByTestId(render.root, 'error-fallback')).toThrow();
    expect(testkit.reports()).toHaveLength(1);

    spy.mockRestore();
  });

  it('sets level appropriately', () => {
    const spy = jest.spyOn(console, 'error');
    spy.mockImplementation(() => undefined);

    // Clear previous reports
    testkit.reset();

    const SessionExpiredComponent = () => {
      throw new SessionExpiredError();
    };

    // Should create warning (reports[0])
    renderer.create(
      <ErrorBoundary renderFallback>
        <SessionExpiredComponent />
      </ErrorBoundary>
    );

    // Should create error (reports[1])
    renderer.create(
      <ErrorBoundary renderFallback>
        <ErrorComponent />
      </ErrorBoundary>
    );

    const reports = testkit.reports();
    expect(reports).toHaveLength(2);
    expect(reports[0].level).toBe('warning');
    expect(reports[1].level).toBe('error');

    spy.mockRestore();
  });

  it('can override fallback components for specific errors', () => {
    const spy = jest.spyOn(console, 'error')
    spy.mockImplementation(() => undefined);

    const SessionExpiredComponent = () => {
      throw new SessionExpiredError();
    };

    const tree = renderer.create(
      <ErrorBoundary
        renderFallback
        errorFallbacks={{
          'SessionExpiredError': <>You are signed out</>,
        }}
        >
        <SessionExpiredComponent />
      </ErrorBoundary>
    ).toJSON();

    expect(tree).toMatchInlineSnapshot(`"You are signed out"`);

    spy.mockRestore();
  });

  it('inits Sentry', () => {
    const initSpy = jest.spyOn(Sentry, 'init');

    act(() => {
      renderer.create(
        <ErrorBoundary sentryDsn='https://examplePublicKey@o0.ingest.sentry.io/0' />
      );
    });

    expect(initSpy).toHaveBeenCalled();
  });

  it('can override Sentry init', () => {
    const initSpy = jest.spyOn(Sentry, 'init');
    const config = {
      dsn: 'https://examplePublicKey@o0.ingest.sentry.io/0',
      enabled: false,
      transport: sentryTransport
    };

    act(() => {
      renderer.create(
        <ErrorBoundary
          sentryInit={config}
        />
      );
    });

    expect(initSpy).toHaveBeenCalledWith(config);
  });
  it('throws if Sentry init is called twice', () => {
    const config = {
      dsn: 'https://examplePublicKey@o0.ingest.sentry.io/0',
      enabled: false,
      transport: sentryTransport
    };

    const instance = renderer.create(
      <ErrorBoundary
        sentryInit={config}
      />
    );

    let caught;
    const saveError = console.error;
    console.error = jest.fn();
    try {
      act(() => instance.update(<ErrorBoundary sentryInit={{dsn: 'foo'}} />))
    } catch (e) {
      caught = e;
    }

    expect(caught).toBe('Sentry.init was already called');
    console.error = saveError;
  });
});
