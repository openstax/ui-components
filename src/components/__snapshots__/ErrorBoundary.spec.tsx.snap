// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary unhandled rejections are captured 1`] = `
<div
  className="sc-eCYdqJ GurhM"
  data-testid="error"
>
  <h3
    className="sc-hKMtZM fYUfXw"
  >
    Uh-oh, there's been a glitch
  </h3>
  We're not quite sure what went wrong. Restart your browser. If this doesn't solve the problem, visit our 
  <a
    href="https://openstax.secure.force.com/help"
    target="_blank"
  >
    Support Center
  </a>
  .
  <div
    className="sc-jqUVSM iQgvmr"
  >
    someuuid
  </div>
</div>
`;

exports[`ErrorBoundary unhandled rejections can be disabled 1`] = `"Content"`;

exports[`ErrorBoundary unhandled rejections does not crash on undefined reasons 1`] = `
<div
  className="sc-eCYdqJ GurhM"
  data-testid="error"
>
  <h3
    className="sc-hKMtZM fYUfXw"
  >
    Uh-oh, there's been a glitch
  </h3>
  We're not quite sure what went wrong. Restart your browser. If this doesn't solve the problem, visit our 
  <a
    href="https://openstax.secure.force.com/help"
    target="_blank"
  >
    Support Center
  </a>
  .
  <div
    className="sc-jqUVSM iQgvmr"
  >
    someuuid
  </div>
</div>
`;
